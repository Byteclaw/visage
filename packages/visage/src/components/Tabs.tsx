/* eslint-disable react/no-array-index-key, jsx-a11y/anchor-is-valid, no-constant-condition */
import { useUniqueId } from '@byteclaw/use-unique-id';
import React, {
  Children,
  KeyboardEventHandler,
  MouseEventHandler,
  ReactElement,
  ReactNode,
  useCallback,
  useRef,
  useState,
  useMemo,
  FocusEvent,
  cloneElement,
} from 'react';
import { ExtractVisageComponentProps } from '@byteclaw/visage-core';
import { createComponent } from '../core';
import { Box } from './Box';
import { Flex } from './Flex';

const TabList = createComponent('div', {
  displayName: 'TabList',
  defaultStyles: {
    boxShadow: 'none',
    display: 'flex',
    m: 0,
    p: 0,
  },
});

const TabNavigatonButton = createComponent('button', {
  displayName: 'TabNavigatonButton',
  defaultStyles: {
    backgroundColor: 'transparent',
    borderColor: 'transparent',
    borderStyle: 'solid',
    borderBottomWidth: 2,
    fontFamily: 'body',
    color: 'lightShadesText',
    cursor: 'pointer',
    fontSize: 1,
    lineHeight: 1,
    outline: 'none',
    py: 1,
    pl: 0,
    pr: 2,
    mr: 1,
    textDecoration: 'none',
    '&[aria-selected="true"]': {
      color: 'primary',
      borderBottomColor: 'primary',
      fontWeight: 'bold',
    },
    '&:focus': {
      borderBottomColor: 'darkAccent',
    },
    '&[disabled]': {
      cursor: 'not-allowed',
      opacity: 0.3,
    },
  },
});

type FlexAllProps = ExtractVisageComponentProps<typeof Flex>;
type TabListAllProps = ExtractVisageComponentProps<typeof TabList>;
type BoxAllProps = ExtractVisageComponentProps<typeof Box>;
type BoxProps = Pick<
  BoxAllProps,
  Exclude<keyof BoxAllProps, 'children' | 'onKeyDown' | 'onClick'>
>;

interface TabProps extends BoxProps {
  children?: ReactNode | (() => ReactNode);
  disabled?: boolean;
  label: string | ReactElement;
  onClick?: MouseEventHandler<HTMLButtonElement>;
  onKeyDown?: KeyboardEventHandler<HTMLButtonElement>;
  selected?: boolean;
}

export function Tab({
  children,
  selected,
  // strip
  label,
  onClick,
  onKeyDown,
  // end strip
  ...restProps
}: TabProps) {
  return (
    <Box {...restProps}>
      {typeof children === 'function'
        ? selected
          ? children()
          : null
        : children}
    </Box>
  );
}

interface TabsProps extends FlexAllProps {
  children: ReactElement<TabProps>[];
  /**
   * Unique id used to generate references between tabs (accessibility)
   */
  id?: string;
  selected?: number;
  tabListProps?: TabListAllProps;
}

export function Tabs({
  children,
  id: outerId,
  selected = 0,
  tabListProps,
  ...restProps
}: TabsProps) {
  const idTemplate = useUniqueId();
  const id = useMemo(() => {
    return outerId || idTemplate;
  }, [outerId, idTemplate]);
  const childrenArray = Children.toArray(children);
  const tabsLabel = childrenArray.map(c => c.props.label);

  const selectedRef = useRef(selected);
  const focusedTabRef = useRef<HTMLButtonElement | null>(null);
  const [selectedTab, selectTab] = useState(selected);
  const onClick: MouseEventHandler<HTMLButtonElement> = useCallback(e => {
    const index = Number(e.currentTarget.dataset.index);

    if (e.currentTarget.getAttribute('aria-disabled') === 'true') {
      return;
    }

    if (childrenArray[index]!.props.onClick) {
      e.persist();

      childrenArray[index]!.props.onClick!(e);
    }

    if (e.defaultPrevented) {
      return;
    }

    e.preventDefault();

    selectTab(index);
  }, []);
  const onTabFocus = useCallback((e: FocusEvent<HTMLButtonElement>) => {
    focusedTabRef.current = e.currentTarget;
  }, []);
  const onKeyDown: KeyboardEventHandler<HTMLButtonElement> = useCallback(e => {
    const index = Number(e.currentTarget.dataset.index);

    if (e.currentTarget.disabled) {
      return;
    }

    switch (e.key) {
      case 'Enter': {
        if (childrenArray[index]!.props.onKeyDown) {
          e.persist();

          childrenArray[index]!.props.onKeyDown!(e);
        }

        if (e.defaultPrevented) {
          return;
        }

        e.preventDefault();

        selectTab(index);
        break;
      }
      case 'ArrowLeft':
      case 'ArrowRight': {
        e.preventDefault();

        let tab: HTMLButtonElement | null = focusedTabRef.current![
          e.key === 'ArrowLeft'
            ? 'previousElementSibling'
            : 'nextElementSibling'
        ] as HTMLButtonElement | null;

        do {
          if (tab) {
            if (tab.disabled) {
              // go to next sibling
              tab = tab[
                e.key === 'ArrowLeft'
                  ? 'previousElementSibling'
                  : 'nextElementSibling'
              ] as HTMLButtonElement | null;
            } else {
              break;
            }
          } else {
            // there is no next tab, find the first children of a parent
            tab = focusedTabRef.current!.parentElement![
              e.key === 'ArrowLeft' ? 'lastElementChild' : 'firstElementChild'
            ] as HTMLButtonElement | null;
          }
        } while (true);

        tab.focus();
        break;
      }
      case 'End':
      case 'Home': {
        e.preventDefault();

        let tab: HTMLButtonElement | null = focusedTabRef.current!
          .parentElement![
          e.key === 'End' ? 'lastElementChild' : 'firstElementChild'
        ] as HTMLButtonElement | null;

        do {
          if (tab) {
            if (tab.disabled) {
              tab = tab[
                e.key === 'End'
                  ? 'previousElementSibling'
                  : 'nextElementSibling'
              ] as HTMLButtonElement | null;
            } else {
              break;
            }
          } else {
            tab = focusedTabRef.current!.parentElement!
              .firstElementChild as HTMLButtonElement | null;
          }
        } while (true);

        tab.focus();
        break;
      }
    }
  }, []);

  if (selectedRef.current !== selected) {
    selectedRef.current = selected;
    selectTab(selected);
  }

  return (
    <Flex
      {...restProps}
      styles={{ flexDirection: 'column', width: '100%', ...restProps.styles }}
    >
      <TabList role="tablist" {...tabListProps}>
        {tabsLabel.map((tabLabel, i) => {
          const isSelected = selectedTab === i;
          const isDisabled = !!childrenArray[i]!.props.disabled;

          return (
            <TabNavigatonButton
              id={`${id}-tab-label-${i}`}
              aria-controls={`${id}-tab-panel-${i}`}
              aria-selected={isSelected}
              disabled={isDisabled}
              data-index={i}
              key={i}
              onClick={onClick}
              onFocus={onTabFocus}
              onKeyDown={onKeyDown}
              role="tab"
              tabIndex={isSelected ? 0 : -1}
              type="button"
            >
              {tabLabel}
            </TabNavigatonButton>
          );
        })}
      </TabList>
      {childrenArray.map((tab, i) =>
        cloneElement(tab, {
          'aria-labelledby': `${id}-tab-label-${i}`,
          id: `${id}-tab-panel-${i}`,
          hidden: selectedTab !== i,
          role: 'tabpanel',
          selected: selectedTab === i,
          tabIndex: 0,
        }),
      )}
    </Flex>
  );
}
