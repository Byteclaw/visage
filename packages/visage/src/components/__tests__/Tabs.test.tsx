import React from 'react';
import { fireEvent } from '@testing-library/react';
import { Tab, Tabs } from '../Tabs';
import { createTestTheme, render } from './render';

describe('Tabs', () => {
  it('renders correctly', () => {
    const { asFragment, getByText, getByTestId } = render(
      <Tabs>
        <Tab label="Tab 1" data-testid="content1">
          Content 1
        </Tab>
        <Tab disabled label="Tab 2" data-testid="content2">
          Content 2
        </Tab>
        <Tab label="Tab 3" data-testid="content3">
          Content 3
        </Tab>
      </Tabs>,
    );

    expect(asFragment()).toMatchSnapshot();

    expect(getByText('Tab 1')).toHaveAttribute('aria-selected', 'true');
    expect(getByTestId('content1')).not.toHaveAttribute('hidden');
  });

  it('works correctly with mouse interaction', () => {
    const { getByText, getByTestId } = render(
      <Tabs>
        <Tab label="Tab 1" data-testid="content1">
          Content 1
        </Tab>
        <Tab disabled label="Tab 2" data-testid="content2">
          Content 2
        </Tab>
        <Tab label="Tab 3" data-testid="content3">
          Content 3
        </Tab>
      </Tabs>,
    );

    expect(getByText('Tab 1')).toHaveAttribute('aria-selected', 'true');
    expect(getByTestId('content1')).not.toHaveAttribute('hidden');

    // click on tab 2 (disabled) does not cause anything
    fireEvent.click(getByText('Tab 2'));

    expect(getByText('Tab 1')).toHaveAttribute('aria-selected', 'true');
    expect(getByTestId('content1')).not.toHaveAttribute('hidden');

    // click on tab 2 (disabled) does not cause anything
    fireEvent.click(getByText('Tab 3'));

    expect(getByText('Tab 3')).toHaveAttribute('aria-selected', 'true');
    expect(getByTestId('content3')).not.toHaveAttribute('hidden');
  });

  it('works correctly with keyboard interaction', () => {
    const { getByText, getByTestId } = render(
      <Tabs>
        <Tab label="Tab 1" data-testid="content1">
          Content 1
        </Tab>
        <Tab disabled label="Tab 2" data-testid="content2">
          Content 2
        </Tab>
        <Tab label="Tab 3" data-testid="content3">
          Content 3
        </Tab>
      </Tabs>,
    );

    expect(getByText('Tab 1')).toHaveAttribute('aria-selected', 'true');
    expect(getByTestId('content1')).not.toHaveAttribute('hidden');

    // focus first button
    fireEvent.focus(getByText('Tab 1'));

    expect(getByText('Tab 1')).toHaveAttribute('aria-selected', 'true');
    expect(getByTestId('content1')).not.toHaveAttribute('hidden');

    // now move to the right, move focus to Tab 3
    fireEvent.keyDown(getByText('Tab 1'), { key: 'ArrowRight' });

    expect(getByText('Tab 3')).toHaveFocus();

    // now press enter
    fireEvent.keyDown(getByText('Tab 3'), { key: 'Enter' });

    expect(getByText('Tab 3')).toHaveAttribute('aria-selected', 'true');
    expect(getByTestId('content3')).not.toHaveAttribute('hidden');

    // now press Home, moves focus to Tab 1
    fireEvent.keyDown(getByText('Tab 3'), { key: 'Home' });

    expect(getByText('Tab 1')).toHaveFocus();

    // now press space to active tab
    fireEvent.keyDown(getByText('Tab 1'), { key: ' ' });

    expect(getByText('Tab 1')).toHaveAttribute('aria-selected', 'true');
    expect(getByTestId('content1')).not.toHaveAttribute('hidden');

    // now press End to move focus to Tab 3
    fireEvent.keyDown(getByText('Tab 1'), { key: 'End' });

    expect(getByText('Tab 3')).toHaveFocus();

    // now press arrow right, to focus Tab 1
    fireEvent.keyDown(getByText('Tab 1'), { key: 'ArrowRight' });

    expect(getByText('Tab 1')).toHaveFocus();

    // now press arrow left, to focus Tab 3
    fireEvent.keyDown(getByText('Tab 1'), { key: 'ArrowLeft' });

    expect(getByText('Tab 3')).toHaveFocus();
  });

  it('is extendable by Tabs face', () => {
    const { asFragment } = render(
      <Tabs>
        <Tab label="Tab 1" data-testid="content1">
          Content 1
        </Tab>
        <Tab disabled label="Tab 2" data-testid="content2">
          Content 2
        </Tab>
        <Tab label="Tab 3" data-testid="content3">
          Content 3
        </Tab>
      </Tabs>,
      {
        ds: {
          theme: createTestTheme({
            Tabs: {
              backgroundColor: 'pink',
            },
          }),
        },
      },
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('is extendable by TabList face', () => {
    const { asFragment } = render(
      <Tabs>
        <Tab label="Tab 1" data-testid="content1">
          Content 1
        </Tab>
        <Tab disabled label="Tab 2" data-testid="content2">
          Content 2
        </Tab>
        <Tab label="Tab 3" data-testid="content3">
          Content 3
        </Tab>
      </Tabs>,
      {
        ds: {
          theme: createTestTheme({
            TabList: {
              borderColor: 'pink',
            },
          }),
        },
      },
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('is extendable by TabNavigationButton face', () => {
    const { asFragment } = render(
      <Tabs>
        <Tab label="Tab 1" data-testid="content1">
          Content 1
        </Tab>
        <Tab disabled label="Tab 2" data-testid="content2">
          Content 2
        </Tab>
        <Tab label="Tab 3" data-testid="content3">
          Content 3
        </Tab>
      </Tabs>,
      {
        ds: {
          theme: createTestTheme({
            TabNavigationButton: {
              color: 'pink',
            },
          }),
        },
      },
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
