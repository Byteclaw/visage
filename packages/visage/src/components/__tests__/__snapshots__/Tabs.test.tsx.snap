// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs is extendable by TabList face 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

.emotion-0 {
  background-color: rgba(0,0,0,0);
  border-color: rgba(0,0,0,0);
  border-style: solid;
  border-width: 0;
  border-bottom-width: 2px;
  font-family: sans-serif;
  color: rgb(255,255,255);
  cursor: pointer;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-size: 26px;
  line-height: 48px;
  outline: none;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 0px;
  padding-right: 16px;
  margin: 0px;
  margin-right: 8px;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: border-color 150ms ease-out;
  transition: border-color 150ms ease-out;
  will-change: border-color;
}

.emotion-0[aria-selected="true"] {
  color: rgb(173,216,230);
  border-bottom-color: rgb(173,216,230);
  font-weight: bold;
}

.emotion-0:focus {
  border-bottom-color: rgb(255,255,255);
}

.emotion-0[disabled] {
  cursor: default;
  opacity: 0.3;
}

.emotion-3 {
  border-bottom-width: 1px;
  border-bottom-color: rgb(184,184,184);
  border-bottom-style: solid;
  box-shadow: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin: 0px;
  margin-bottom: 8px;
  overflow-x: auto;
  padding: 0px;
  width: 100%;
  -ms-overflow-style: none;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
  border-color: rgb(255,192,203);
}

.emotion-3::-webkit-scrollbar {
  display: none;
}

<div
    class="emotion-7"
  >
    <div
      class="emotion-3"
      role="tablist"
    >
      <button
        aria-controls="tabs-1-tab-panel-0"
        aria-selected="true"
        class="emotion-0"
        data-index="0"
        id="tabs-1-tab-label-0"
        role="tab"
        tabindex="0"
        type="button"
      >
        Tab 1
      </button>
      <button
        aria-controls="tabs-1-tab-panel-1"
        aria-selected="false"
        class="emotion-0"
        data-index="1"
        disabled=""
        id="tabs-1-tab-label-1"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Tab 2
      </button>
      <button
        aria-controls="tabs-1-tab-panel-2"
        aria-selected="false"
        class="emotion-0"
        data-index="2"
        id="tabs-1-tab-label-2"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Tab 3
      </button>
    </div>
    <div
      aria-labelledby="tabs-1-tab-label-0"
      class="emotion-4"
      data-testid="content1"
      id="tabs-1-tab-panel-0"
      role="tabpanel"
      tabindex="0"
    >
      Content 1
    </div>
    <div
      aria-labelledby="tabs-1-tab-label-1"
      class="emotion-4"
      data-testid="content2"
      disabled=""
      hidden=""
      id="tabs-1-tab-panel-1"
      role="tabpanel"
      tabindex="0"
    >
      Content 2
    </div>
    <div
      aria-labelledby="tabs-1-tab-label-2"
      class="emotion-4"
      data-testid="content3"
      hidden=""
      id="tabs-1-tab-panel-2"
      role="tabpanel"
      tabindex="0"
    >
      Content 3
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tabs is extendable by TabNavigationButton face 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

.emotion-3 {
  border-bottom-width: 1px;
  border-bottom-color: rgb(184,184,184);
  border-bottom-style: solid;
  box-shadow: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin: 0px;
  margin-bottom: 8px;
  overflow-x: auto;
  padding: 0px;
  width: 100%;
  -ms-overflow-style: none;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
}

.emotion-3::-webkit-scrollbar {
  display: none;
}

.emotion-0 {
  background-color: rgba(0,0,0,0);
  border-color: rgba(0,0,0,0);
  border-style: solid;
  border-width: 0;
  border-bottom-width: 2px;
  font-family: sans-serif;
  color: rgb(255,192,203);
  cursor: pointer;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-size: 26px;
  line-height: 48px;
  outline: none;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 0px;
  padding-right: 16px;
  margin: 0px;
  margin-right: 8px;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: border-color 150ms ease-out;
  transition: border-color 150ms ease-out;
  will-change: border-color;
}

.emotion-0[aria-selected="true"] {
  color: rgb(173,216,230);
  border-bottom-color: rgb(173,216,230);
  font-weight: bold;
}

.emotion-0:focus {
  border-bottom-color: rgb(255,255,255);
}

.emotion-0[disabled] {
  cursor: default;
  opacity: 0.3;
}

<div
    class="emotion-7"
  >
    <div
      class="emotion-3"
      role="tablist"
    >
      <button
        aria-controls="tabs-1-tab-panel-0"
        aria-selected="true"
        class="emotion-0"
        data-index="0"
        id="tabs-1-tab-label-0"
        role="tab"
        tabindex="0"
        type="button"
      >
        Tab 1
      </button>
      <button
        aria-controls="tabs-1-tab-panel-1"
        aria-selected="false"
        class="emotion-0"
        data-index="1"
        disabled=""
        id="tabs-1-tab-label-1"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Tab 2
      </button>
      <button
        aria-controls="tabs-1-tab-panel-2"
        aria-selected="false"
        class="emotion-0"
        data-index="2"
        id="tabs-1-tab-label-2"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Tab 3
      </button>
    </div>
    <div
      aria-labelledby="tabs-1-tab-label-0"
      class="emotion-4"
      data-testid="content1"
      id="tabs-1-tab-panel-0"
      role="tabpanel"
      tabindex="0"
    >
      Content 1
    </div>
    <div
      aria-labelledby="tabs-1-tab-label-1"
      class="emotion-4"
      data-testid="content2"
      disabled=""
      hidden=""
      id="tabs-1-tab-panel-1"
      role="tabpanel"
      tabindex="0"
    >
      Content 2
    </div>
    <div
      aria-labelledby="tabs-1-tab-label-2"
      class="emotion-4"
      data-testid="content3"
      hidden=""
      id="tabs-1-tab-panel-2"
      role="tabpanel"
      tabindex="0"
    >
      Content 3
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tabs is extendable by Tabs face 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-3 {
  border-bottom-width: 1px;
  border-bottom-color: rgb(184,184,184);
  border-bottom-style: solid;
  box-shadow: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin: 0px;
  margin-bottom: 8px;
  overflow-x: auto;
  padding: 0px;
  width: 100%;
  -ms-overflow-style: none;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
}

.emotion-3::-webkit-scrollbar {
  display: none;
}

.emotion-0 {
  background-color: rgba(0,0,0,0);
  border-color: rgba(0,0,0,0);
  border-style: solid;
  border-width: 0;
  border-bottom-width: 2px;
  font-family: sans-serif;
  color: rgb(255,255,255);
  cursor: pointer;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-size: 26px;
  line-height: 48px;
  outline: none;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 0px;
  padding-right: 16px;
  margin: 0px;
  margin-right: 8px;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: border-color 150ms ease-out;
  transition: border-color 150ms ease-out;
  will-change: border-color;
}

.emotion-0[aria-selected="true"] {
  color: rgb(173,216,230);
  border-bottom-color: rgb(173,216,230);
  font-weight: bold;
}

.emotion-0:focus {
  border-bottom-color: rgb(255,255,255);
}

.emotion-0[disabled] {
  cursor: default;
  opacity: 0.3;
}

.emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  background-color: rgb(255,192,203);
}

<div
    class="emotion-7"
  >
    <div
      class="emotion-3"
      role="tablist"
    >
      <button
        aria-controls="tabs-1-tab-panel-0"
        aria-selected="true"
        class="emotion-0"
        data-index="0"
        id="tabs-1-tab-label-0"
        role="tab"
        tabindex="0"
        type="button"
      >
        Tab 1
      </button>
      <button
        aria-controls="tabs-1-tab-panel-1"
        aria-selected="false"
        class="emotion-0"
        data-index="1"
        disabled=""
        id="tabs-1-tab-label-1"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Tab 2
      </button>
      <button
        aria-controls="tabs-1-tab-panel-2"
        aria-selected="false"
        class="emotion-0"
        data-index="2"
        id="tabs-1-tab-label-2"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Tab 3
      </button>
    </div>
    <div
      aria-labelledby="tabs-1-tab-label-0"
      class="emotion-4"
      data-testid="content1"
      id="tabs-1-tab-panel-0"
      role="tabpanel"
      tabindex="0"
    >
      Content 1
    </div>
    <div
      aria-labelledby="tabs-1-tab-label-1"
      class="emotion-4"
      data-testid="content2"
      disabled=""
      hidden=""
      id="tabs-1-tab-panel-1"
      role="tabpanel"
      tabindex="0"
    >
      Content 2
    </div>
    <div
      aria-labelledby="tabs-1-tab-label-2"
      class="emotion-4"
      data-testid="content3"
      hidden=""
      id="tabs-1-tab-panel-2"
      role="tabpanel"
      tabindex="0"
    >
      Content 3
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tabs renders correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

.emotion-3 {
  border-bottom-width: 1px;
  border-bottom-color: rgb(184,184,184);
  border-bottom-style: solid;
  box-shadow: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin: 0px;
  margin-bottom: 8px;
  overflow-x: auto;
  padding: 0px;
  width: 100%;
  -ms-overflow-style: none;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
}

.emotion-3::-webkit-scrollbar {
  display: none;
}

.emotion-0 {
  background-color: rgba(0,0,0,0);
  border-color: rgba(0,0,0,0);
  border-style: solid;
  border-width: 0;
  border-bottom-width: 2px;
  font-family: sans-serif;
  color: rgb(255,255,255);
  cursor: pointer;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-size: 26px;
  line-height: 48px;
  outline: none;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 0px;
  padding-right: 16px;
  margin: 0px;
  margin-right: 8px;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: border-color 150ms ease-out;
  transition: border-color 150ms ease-out;
  will-change: border-color;
}

.emotion-0[aria-selected="true"] {
  color: rgb(173,216,230);
  border-bottom-color: rgb(173,216,230);
  font-weight: bold;
}

.emotion-0:focus {
  border-bottom-color: rgb(255,255,255);
}

.emotion-0[disabled] {
  cursor: default;
  opacity: 0.3;
}

<div
    class="emotion-7"
  >
    <div
      class="emotion-3"
      role="tablist"
    >
      <button
        aria-controls="tabs-1-tab-panel-0"
        aria-selected="true"
        class="emotion-0"
        data-index="0"
        id="tabs-1-tab-label-0"
        role="tab"
        tabindex="0"
        type="button"
      >
        Tab 1
      </button>
      <button
        aria-controls="tabs-1-tab-panel-1"
        aria-selected="false"
        class="emotion-0"
        data-index="1"
        disabled=""
        id="tabs-1-tab-label-1"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Tab 2
      </button>
      <button
        aria-controls="tabs-1-tab-panel-2"
        aria-selected="false"
        class="emotion-0"
        data-index="2"
        id="tabs-1-tab-label-2"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Tab 3
      </button>
    </div>
    <div
      aria-labelledby="tabs-1-tab-label-0"
      class="emotion-4"
      data-testid="content1"
      id="tabs-1-tab-panel-0"
      role="tabpanel"
      tabindex="0"
    >
      Content 1
    </div>
    <div
      aria-labelledby="tabs-1-tab-label-1"
      class="emotion-4"
      data-testid="content2"
      disabled=""
      hidden=""
      id="tabs-1-tab-panel-1"
      role="tabpanel"
      tabindex="0"
    >
      Content 2
    </div>
    <div
      aria-labelledby="tabs-1-tab-label-2"
      class="emotion-4"
      data-testid="content3"
      hidden=""
      id="tabs-1-tab-panel-2"
      role="tabpanel"
      tabindex="0"
    >
      Content 3
    </div>
  </div>
</DocumentFragment>
`;
